### ISSUE TITLE
User data socket closes after 1 hour of inactivity

### ISSUE THREAD

**Issue by tiagosiebler** [tiagosiebler/binance#41]
Hello! Loving this implementation so far! I've been playing with the user data websocket and I've noticed that after 1 hour of inactivity it closes with code 1006. This is the full object passed through the onclose event of the socket:
```
{
	"target": {
		"domain": null,
		"_events": {
			"error": [null, null, null],
			"close": [null, null]
		},
		"_eventsCount": 3,
		"readyState": 3,
		"bytesReceived": 87667,
		"extensions": {},
		"protocol": "",
		"_binaryType": "nodebuffer",
		"_closeFrameReceived": false,
		"_closeFrameSent": false,
		"_closeMessage": "",
		"_closeTimer": null,
		"_finalized": true,
		"_closeCode": 1006,
		"_receiver": {
			"_binaryType": "nodebuffer",
			"_extensions": null,
			"_maxPayload": 0,
			"_bufferedBytes": 0,
			"_buffers": null,
			"_compressed": false,
			"_payloadLength": 6882,
			"_fragmented": 0,
			"_masked": false,
			"_fin": true,
			"_mask": null,
			"_opcode": 1,
			"_totalPayloadLength": 0,
			"_messageLength": 0,
			"_fragments": null,
			"_cleanupCallback": null,
			"_hadError": false,
			"_dead": true,
			"_loop": false,
			"onmessage": null,
			"onclose": null,
			"onerror": null,
			"onping": null,
			"onpong": null,
			"_state": 0
		},
		"_sender": null,
		"_socket": null,
		"_ultron": null,
		"protocolVersion": 13,
		"_isServer": false,
		"url": "wss://stream.binance.com:9443/ws/somethingosmething",
		"_req": null
	},
	"type": "close",
	"wasClean": false,
	"reason": "",
	"code": 1006
}
```

I've reduced the keep alive timer to 30000 ms, in case that may be a factor, but it doesn't seem related as the timing remains 1 hour after the last activity. Hope the onclose event object is useful in diagnosing this. This is with the latest version (1.3.3).

**Comment by aarongarvey** [comment]
Thanks for the details.  I've been doing some web socket longevity testing
myself hoping to identify the different failure conditions so I can
implement some reconnect functionality.  Should hopefully have that and a
few other websocket related improvements done within the next week or two.

**Comment by tiagosiebler** [comment]
@aarongarvey thank you, I look forward to it!

**Comment by brumack** [comment]
I've been playing with this myself and haven't yet found a solution. I tried passing the binanceRest object through to setupWebSocket in the BinanceWS class so it had access to the startUserDataStream method along with the auth credentials, and then call startUserDataStream on 'close' but came up with errors. It was meant to just be a bandaid but no luck so far. I'm logging the payload from the keepalive method to the console just to be sure its coming back with a response.

**Comment by brumack** [comment]
Ok turns out my method above does actually work. Pass the binanceRest object to onUserData in ws.js, and then pass it from there to ._setupWebSocket. Then, set up an event handler for 'close' and call the startUserDataStream method of the binanceRest object. Lmk if you'd like me to submit a PR.

**Comment by laurynas-karvelis** [comment]
Any update on this socket issue? I've noticed same behaviour, reduced keepalive timeout to 30mins, but it seems socket gets closed and the attached listener never gets called anymore.

**Comment by tiagosiebler** [comment]
@brumack did you manage to patch this up somehow? A thought I had was simply running binanceWS.onUserData again to start a new socket, when the last one closes like this.

**Comment by brumack** [comment]
@tiagosiebler still haven't found a solution :(

**Comment by laurynas-karvelis** [comment]
Hi all, I've submitted a PR that works around all disconnects (public and onUserData socket) and reconnects as soon as possible by also emitting an event just before a `reconnect` and after `connection` is established. [Here's the URL](https://github.com/aarongarvey/binance/pull/49)

**Comment by tiagosiebler** [comment]
Fixed in #122 - check the example here for new syntax:
https://github.com/tiagosiebler/binance/blob/v2.0.0/beta/examples/ws-userdata.ts
